### PR TITLE
BUG: Fix histogram bins="auto" for data with little variance

### LINKED ISSUE #9327: 'auto' selection of histogram bins fails in some cases
This is an odd corner case, but there are times using auto to select the number of bins gives only a single bin.

    from pylab import *
    data = zeros(1000)
    data[:100] = 10
    np.histogram(data, bins='auto')

Output is a single bin

     (array([1000]), array([  0.,  10.]))

This must be the result of the sturges/fd estimators since (according to the documentation) auto just takes the max of these two.  It seems like this should fall back to the default number of bins (10 I think), or perhaps sqrt.

### PR DESCRIPTION
**Pull request by nayyarv**
As brought up in #9327, data with very low variance will result in 0 IQR and thus 1 bin being chosen, which is not optimal.

In situations when the IQR is 0, using an estimator dependent on a measure of variance is not a great idea. I think we should just fall back to using a simple estimator such as sturges which is purely dependent on size and given it's use in R, is still a reasonable default to choose. I could also be prompted to choose between the rice estimator and sturges since the rice estimator will be larger for large data sizes.

@madphysicist suggested 
>  Instead, we could modify _hist_bin_fd to check if the IQR is zero because the data is shaped unfortunately vs actually having only one value. For example, if IQR is zero, use np.unique to see how many distinct values there are and use the min/max values instead

I'm not super keen on the above - I think using a standard method should do what the method intended so we're consistent with literature. Similarly, using `np.unique` is likely a fools errand since 

```python
a = np.zeros(1000)
a[:248] = -1 * np.random.random(248)
a[751:] =  np.random.random(248)
```
Would have ~500 unique values and still suffer the same problem brought up in #9327. 

I think it's fine not being able to suggest the right number of bins for non-normal and pathological data, since this is not easy to do in a general way, and the estimators are based on some assumptions which don't hold and there should be no expectation for good performance in such situations.

In this case, we really want to choose the `max(fd_bins, sturges_bins)` and by doing `min(fd_bw, sturges_bw)` we are not doing that when the IQR is 0. This PR makes the behaviour consistent with the original intention.

Fixes #9327

### CODE REVIEW

File: numpy/lib/histograms.py [hunk 1]
```diff
@@ -184,10 +191,12 @@ def _hist_bin_auto(x):
     --------
     _hist_bin_fd, _hist_bin_sturges
     """
-    # There is no need to check for zero here. If ptp is, so is IQR and
-    # vice versa. Either both are zero or neither one is.
-    return min(_hist_bin_fd(x), _hist_bin_sturges(x))
-
+    fd_bw = _hist_bin_fd(x)
+    if fd_bw:
+        return min(_hist_bin_fd(x), _hist_bin_sturges(x))
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: No point recalculating fd_bw here.

Might be tidier to compute `fd_sturges` once before both paths too.

Comment (nayyarv) [reply numpy/lib/histograms.py]: Oh dear, that was my initial plan, but looks like I forgot when I put the commit together haha.

File: numpy/lib/histograms.py [hunk 2]
```diff
@@ -163,12 +163,19 @@ def _hist_bin_fd(x):
...
     The FD estimator is usually the most robust method, but its width
-    estimate tends to be too large for small `x`. The Sturges estimator
-    is quite good for small (<1000) datasets and is the default in the R
-    language. This method gives good off the shelf behaviour.
+    estimate tends to be too large for small `x` and bad for data with limited
+    variance. The Sturges estimator is quite good for small (<1000) datasets
+    and is the default in the R language. This method gives good off the shelf
+    behaviour.
+
+    if there is limited variance, the IQR will be 0 which will result in
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Capital I

File: numpy/lib/histograms.py [hunk 3]
```diff
@@ -163,12 +163,19 @@ def _hist_bin_fd(x):
...
+    1 bin being chosen which may not be optimal. If the IQR is 0, it's
+    unlikely any variance based estimators will be of use, so we revert to
+    a purely size based calculation
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Missing `.`

File: numpy/lib/histograms.py (hunk 2 above)
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Perhaps add `.. versionchanged:: 1.15.0` the line above here too

File: numpy/lib/tests/test_histograms.py [hunk 4]
```diff
@@ -430,6 +430,24 @@ def test_novariance(self):
...
+    def test_limited_variance(self):
+        """
+        check in the situation that we have some variance and should return
+        the sturges value and not the fd value (which should be 0)
+        """
+        lim_var_data = np.ones(1000)
+        lim_var_data[:3] = 0
+        lim_var_data[-4:] = 100
+
+        a, b = np.histogram(lim_var_data, 'auto')
```
Comment (eric-wieser) [inline numpy/lib/tests/test_histograms.py]: May as well just test `histogram_bin_edges` here now.

**Comment by nayyarv** [comment]
Updated tests to suggestion and squashed commits.

File: numpy/lib/histograms.py [hunk 5]
```diff
@@ -184,10 +192,13 @@ def _hist_bin_auto(x):
     --------
     _hist_bin_fd, _hist_bin_sturges
     """
-    # There is no need to check for zero here. If ptp is, so is IQR and
```
Comment (madphysicist) [inline numpy/lib/histograms.py]: I see the error of my ways now :)

File: numpy/lib/histograms.py [hunk 6]
```diff
@@ -436,7 +447,7 @@ def histogram_bin_edges(a, bins=10, range=None, weights=None):
     below, :math:`h` is the binwidth and :math:`n_h` is the number of
     bins. All estimators that compute bin counts are recast to bin width
     using the `ptp` of the data. The final bin count is obtained from
-    ``np.round(np.ceil(range / h))`.
+    `np.round(np.ceil(range / h))`.
```
Comment (madphysicist) [inline numpy/lib/histograms.py]: I think the correct version is to have two ticks on both sides. That's the sphinx symbol for a code block, vs the single tick which is "format how you see fit".

**Comment by madphysicist** [comment]
Sorry I didn't get a chance to look at this earlier. It looks great. This is really the simplest and most elegant possible solution.

File: numpy/lib/histograms.py [hunk 7]
```diff
@@ -163,12 +163,20 @@ def _hist_bin_fd(x):
...
-    is quite good for small (<1000) datasets and is the default in the R
-    language. This method gives good off the shelf behaviour.
+    estimate tends to be too large for small `x` and bad for data with limited
+    variance. The Sturges estimator is quite good for small (<1000) datasets
+    and is the default in the R language. This method gives good off the shelf
+    behaviour.
+
+    .. versionchanged:: 1.15.0
+    If there is limited variance, the IQR will be 0 which will result in
+    1 bin being chosen which may not be optimal. If the IQR is 0, it's
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Perhaps this should say that "when IQR is 0, the bin width from the FD estimator is 0, so instead ...". "1 bin being chosen" is arguably a bug in `histogram_bin_edges`

Comment (eric-wieser) [reply numpy/lib/histograms.py]: To be clear, `1` is chosen not due to any hope that it is optimal, but because it is a more useful choice than `np.inf`.

**Review by eric-wieser — COMMENTED** [review]
Fix looks correct, not super clear on the docs just yet

File: numpy/lib/histograms.py (hunk 6 above)
Comment (nayyarv) [reply numpy/lib/histograms.py]: Oh dear, this entire docstring uses primarily single backticks

Comment (nayyarv) [reply numpy/lib/histograms.py]: Yeah you're right, the entire docstring primarily uses single backticks though.

File: numpy/lib/histograms.py (hunk 7 above)
Comment (nayyarv) [reply numpy/lib/histograms.py]: when no data is passed in or the data has no variance, 1 bin is correct, just not always.

**Comment by nayyarv** [comment]
Thanks @madphysicist @eric-wieser.

I've updated the docs and squashed to 1 commit.

File: numpy/lib/histograms.py [hunk 8]
```diff
@@ -163,12 +163,22 @@ def _hist_bin_fd(x):
...
-    is quite good for small (<1000) datasets and is the default in the R
-    language. This method gives good off the shelf behaviour.
+    estimate tends to be too large for small `x` and bad for data with limited
+    variance. The Sturges estimator is quite good for small (<1000) datasets
+    and is the default in the R language. This method gives good off the shelf
+    behaviour.
+
+    .. versionchanged:: 1.15.0
+    If there is limited variance, the IQR can be 0 which will result in the
+    FD bandwitdh being 0 too. This is not a valid bandwidth and so
```
Comment (eric-wieser) [inline numpy/lib/histograms.py]: Typo: bandwidth. I think bin width is clearer though

**Comment by nayyarv** [comment]
Oops, forgot to push my changes after last comment, but looks like @eric-wieser got to them. Again resquashed.

**Review by eric-wieser — APPROVED** [review]

**Review by madphysicist — APPROVED** [review]
This is a neat feature.

**Comment by nayyarv** [comment]
Thanks!

Am I supposed to do something now?

**Comment by eric-wieser** [comment]
Thanks @nayyarv! Was just leaving this for any more comments, but it seems uncontroversial.

**Pull request merged**